packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts

```diff
@@ -0,0 +1,153 @@
+            this.authenticationProviders.set(registry, provider);
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return this.authenticationProviders.get(registry)!;
+    }
+
+    public async addTrackedRegistry(): Promise<void> {
+        const wizardContext: GenericRegistryV2WizardContext = {
+            registryPrompt: vscode.l10n.t('Registry URL'),
+            registryPromptPlaceholder: vscode.l10n.t('http://localhost:5000'),
+            usernamePrompt: vscode.l10n.t('Registry Username'),
+            secretPrompt: vscode.l10n.t('Registry Password or Personal Access Token'),
+        };
+
+        const wizard = new RegistryWizard(
+            wizardContext,
+            [
+                new GenericRegistryV2WizardPromptStep(),
+                new RegistryWizardUsernamePromptStep(),
+                new RegistryWizardSecretPromptStep(),
+            ],
+            new vscode.CancellationTokenSource().token
+        );
+
+        await wizard.prompt();
+
+        if (!wizardContext.registryUri) {
+            throw new Error('Registry URL is invalid');
+        }
+
+        const registryUriString = wizardContext.registryUri.toString().toLowerCase();
+
+        // store registry url in memento
+        const trackedRegistryStrings = this.extensionContext.globalState.get<string[]>(TrackedRegistriesKey, []);
+        trackedRegistryStrings.push(registryUriString);
+        await this.extensionContext.globalState.update(TrackedRegistriesKey, trackedRegistryStrings);
+
+        // store credentials in auth provider
+        const authProvider = new BasicOAuthProvider(this.extensionContext.globalState, this.extensionContext.secrets, wizardContext.registryUri);
+        await authProvider.storeBasicCredentials(
+            {
+                username: wizardContext.username || '',
+                secret: wizardContext.secret || '',
+            }
+        );
+        this.authenticationProviders.set(registryUriString, authProvider);
+    }
+
+    public async removeTrackedRegistry(registry: GenericV2Registry): Promise<void> {
+        // remove registry url from list of tracked registries
+        const registryUriString = registry.baseUrl.toString();
+        const trackedRegistryStrings = this.extensionContext.globalState.get<string[]>(TrackedRegistriesKey, []);
+        const index = trackedRegistryStrings.findIndex(r => r === registryUriString);
+        if (index !== -1) {
+            trackedRegistryStrings.splice(index, 1);
+            void this.extensionContext.globalState.update(TrackedRegistriesKey, trackedRegistryStrings);
+        }
+
+        // remove credentials from auth provider
+        await this.authenticationProviders.get(registryUriString)?.removeSession();
+        this.authenticationProviders.delete(registryUriString);
+        // TODO: check if the map of auth providers is empty, if so, remove the root from the tree

```

Comment:
Do?

Reply:
mhmm not sure if this is worth, what do you think?

Reply:
Well, is there any way to make the node disappear after a disconnect?

Reply:
no, not currently

Reply:
we'll do this as a separate PR, issue is tracked here https://github.com/microsoft/vscode-docker-extensibility/issues/193